crates/sui-display/src/v2/interpreter.rs

```diff
@@ -0,0 +1,300 @@
+                (V::Slice(slice), _) => {
+                    let Some(value) = Extractor::deserialize_slice(slice, &mut accessors)
+                        .map_err(FormatError::Visitor)?
+                    else {
+                        return Ok(None);
+                    };
+
+                    root = value;
+                }
+
+                // Scalar values do not support accessors.
+                (
+                    V::Address(_)
+                    | V::Bool(_)
+                    | V::String(_)
+                    | V::U8(_)
+                    | V::U16(_)
+                    | V::U32(_)
+                    | V::U64(_)
+                    | V::U128(_)
+                    | V::U256(_),
+                    _,
+                ) => return Ok(None),
+            }
+        }
+
+        // Detect if the value we sliced out was the serialized Move representation of `None` and
+        // convert that to `Ok(None)`, otherwise return the extracted value.
+        Ok(if root.is_none() { None } else { Some(root) })
+    }
+
+    /// Evaluates the contents of an accessor to a value.
+    ///
+    /// Returns `Ok(Some(value))` if the accessor evaluates to a value, otherwise it propagates
+    /// errors or `None` values.
+    async fn eval_accessor(
+        &self,
+        acc: &'a P::Accessor<'a>,
+    ) -> Result<Option<Accessor<'a>>, FormatError> {
+        use Accessor as VA;
+        use P::Accessor as PA;
+
+        Ok(match acc {
+            PA::Field(f) => Some(VA::Field(f.as_str())),
+            PA::Positional(i) => Some(VA::Positional(*i)),
+            PA::Index(chain) => Box::pin(self.eval_chain(chain)).await?.map(VA::Index),
+            PA::DFIndex(chain) => Box::pin(self.eval_chain(chain)).await?.map(VA::DFIndex),
+            PA::DOFIndex(chain) => Box::pin(self.eval_chain(chain)).await?.map(VA::DOFIndex),
+        })
+    }
+
+    /// Evaluate literals to values.
+    ///
+    /// Returns `Ok(Some(value))` if all parts of the literal evaluate to `Ok(Some(value))`,
+    /// otherwise it propagates errors or `None` values.
+    async fn eval_literal(
+        &self,
+        lit: &'a P::Literal<'a>,
+    ) -> Result<Option<Value<'a>>, FormatError> {
+        use Value as V;
+        use P::Literal as L;
+
+        Ok(match lit {
+            L::Address(a) => Some(V::Address(*a)),
+            L::Bool(b) => Some(V::Bool(*b)),
+            L::U8(n) => Some(V::U8(*n)),
+            L::U16(n) => Some(V::U16(*n)),
+            L::U32(n) => Some(V::U32(*n)),
+            L::U64(n) => Some(V::U64(*n)),
+            L::U128(n) => Some(V::U128(*n)),
+            L::U256(n) => Some(V::U256(*n)),
+            L::ByteArray(bs) => Some(V::Bytes(bs.into())),
+            L::String(s) => Some(V::String(s.clone())),
+
+            L::Vector(v) => self.eval_chains(&v.elements).await?.map(|elements| {
+                V::Vector(Vector {
+                    type_: v.type_.as_ref(),
```

Comment:
I see why the `type_` in `Vector` is an `Option<..>` here since we allow `vector<>` syntax, but should we just require `'vector' '<' <type> '>'` instead of `'vector' '<' (<type>,?)? '>'`?

Reply:
We support the following syntax:

- `vector<T>` or `vector<T>[]` for an empty vector.
- `vector[a, b, ..., z]` for a non-empty vector.

It's because of the second case that `type_` is optional, but we can always calculate the type of the vector literal either from its type annotation or from its elements.